field/composite.go

```diff
@@ -117,55 +119,75 @@ func (f *Composite) SetData(data interface{}) error {
 
 // Pack deserialises data held by the receiver (via SetData)
 // into bytes and returns an error on failure.
-func (f *Composite) Pack() ([]byte, error) {
-	packed, err := f.pack()
+func (f *Composite) WriteTo(w io.Writer) (n int, err error) {
+	packed := bytes.NewBuffer([]byte{})
+	err = f.pack(packed)
 	if err != nil {
-		return nil, err
+		return 0, err
+	}
+
+	packedLength, err := f.spec.Pref.EncodeLength(f.spec.Length, packed.Len())
+	if err != nil {
+		return 0, fmt.Errorf("failed to encode length: %v", err)
 	}
 
-	packedLength, err := f.spec.Pref.EncodeLength(f.spec.Length, len(packed))
+	m, err := w.Write(packedLength)
+	if err != nil {
+		return m, fmt.Errorf("writing packed length: %v", err)
+	}
+
+	n += m
+
+	m, err = w.Write(packed.Bytes())
 	if err != nil {
-		return nil, fmt.Errorf("failed to encode length: %v", err)
+		return m, fmt.Errorf("writing packed field: %v", err)
 	}
 
-	return append(packedLength, packed...), nil
+	n += m
+
+	return n, nil
 }
 
-// Unpack takes in a byte array and serializes them into the receiver's
-// subfields. An offset (unit depends on encoding and prefix values) is
-// returned on success. A non-nil error is returned on failure.
-func (f *Composite) Unpack(data []byte) (int, error) {
-	dataLen, err := f.spec.Pref.DecodeLength(f.spec.Length, data)
+// ReadFrom takes in an io.Reader and reads data from it and then serializes
+// them into the receiver's subfields. An offset (unit depends on encoding and
+// prefix values) is returned on success. A non-nil error is returned on
+// failure.
+func (f *Composite) ReadFrom(r io.Reader) (int, error) {
```

Comment:
It seems like we're changing this behavior from:

`Unpack` was requiring the entire message be copied from the network even though we may only need to consume part of it initially? e.g. read message headers
```
AAAABBBBBBCCCCCC (needs to be consumed and written into []byte)
      ^ (the offset we only needed to read from)
```

`ReadFrom` looks to be trying to offer the ability to _only_ consume `AAAABB` if that's all of the message needed for now, but the [Go docs on that interface](https://golang.org/pkg/io/#ReaderFrom) (`ReaderFrom`) mention that the entire message will be consumed still. 

`Composit.ReadFrom` mentions it returns an _offset_ but Go's docs return `n` as the **total number of bytes read**. That behavior breaks the ABI but would still compile and introduce bugs. Also note Go requires it be measured in bytes but this PR's units can vary at runtime. 

Of course this could just be renaming methods so that they conform to Go's stdlib interfaces which allows for interop with lots of packages.

Reply:
Documentation for `Composit.ReadFrom` should be updated. ReadFrom will always return bytes read.

I missed that ReadFrom reads until the EOF.

Reply:
Could we also update the `Field` and `Encoder` interfaces to indicate that the return type in `ReadFrom` is always measured in bytes as well?